[LoopVectorize] Add cost of generating tail-folding mask to the loop

david-arm opened the pull request:
At the moment if we decide to enable tail-folding we do not include the cost of generating the mask per VF. This can mean we make some poor choices of VF, which is definitely true for SVE-enabled AArch64 targets where mask generation for fixed-width vectors is more expensive than for scalable vectors.

New tests added:

  Transforms/LoopVectorize/AArch64/sve-tail-folding-cost.ll
  Transforms/LoopVectorize/RISCV/tail-folding-cost.ll

david-arm commented:
Gentle ping!

huntergr-arm commented on llvm/test/Transforms/LoopVectorize/RISCV/uniform-load-store.ll:
```diff
@@ -497,49 +497,21 @@ define void @conditional_uniform_load(ptr noalias nocapture %a, ptr noalias noca
 ; TF-FIXEDLEN-NEXT:    br label [[FOR_BODY:%.*]]
 ; TF-FIXEDLEN:       for.body:
-; TF-FIXEDLEN-NEXT:    [[IV:%.*]] = phi i64 [ [[BC_RESUME_VAL]], [[SCALAR_PH]] ], [ [[IV_NEXT:%.*]], [[LATCH:%.*]] ]
+; TF-FIXEDLEN-NEXT:    [[IV:%.*]] = phi i64 [ 0, [[ENTRY:%.*]] ], [ [[IV_NEXT:%.*]], [[LATCH:%.*]] ]
```
Seems like this may have inadvertently broken a RVV functionality test? Though the scalable version still works...

david-arm replied:
I think it's because the cost model is now teaching the vectoriser to behave sensibly, i.e. when using tail-folding if we only have fixed-width vectors available it's cheaper to use a scalar loop. That's presumably because for RISCV the compiler expands the active.lane mask into a horrible sequence of instructions. I did think about trying to somehow hack the RUN lines or add loop attributes to force tail-folding for fixed-width vectors, but given there is still testing coverage for the scalable case I figured I'd just leave it. Unless someone from RISCV has a preference - @topperc?

david-arm commented:
Rebase.

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/induction-costs-sve.ll:
The tests have changed because the active lane mask cost is the same for VF=8 and VF=16. As a result the per scalar iteration cost is lower for VF=16.

david-arm commented on llvm/test/Transforms/LoopVectorize/RISCV/uniform-load-store.ll:
```diff
@@ -430,116 +430,41 @@ define void @conditional_uniform_load(ptr noalias nocapture %a, ptr noalias noca
 ;
 ; TF-SCALABLE-LABEL: @conditional_uniform_load(
 ; TF-SCALABLE-NEXT:  entry:
-; TF-SCALABLE-NEXT:    br i1 false, label [[SCALAR_PH:%.*]], label [[VECTOR_PH:%.*]]
```
Both TF-SCALABLE and TF-FIXED now fail to vectorise because the additional lane mask cost has just pushed it over the cost of the scalar loop. At least for the scalable case this may indicate the cost model for get.active.lane.mask calls needs improving?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -5942,6 +5945,35 @@ InstructionCost LoopVectorizationCostModel::computePredInstDiscount(
+        Intrinsic::get_active_lane_mask, RetTy,
+        {PoisonValue::get(IndTy), PoisonValue::get(IndTy)});
+    MaskCost = TTI.getIntrinsicInstrCost(Attrs, CostKind);
+  } else {
+    // This is just a stepvector, added to a splat of the current IV, followed
+    // by a vector comparison with a splat of the trip count. Since the
+    // stepvector is loop invariant it will be hoisted out so we can ignore it.
+    // This just leaves us with an add and an icmp.
+    VectorType *VecTy = VectorType::get(IndTy, VF);
+    MaskCost = TTI.getArithmeticInstrCost(Instruction::Add, VecTy, CostKind);
+    MaskCost += TTI.getCmpSelInstrCost(Instruction::ICmp, VecTy, RetTy,
+                                       ICmpInst::ICMP_ULE, CostKind, nullptr);
+  }
```
This is wrong for tailfolding with EVL, need a special case for this. The cost for data-with-evl tail folding style is the cost of sub i64 + cost of i32 @llvm.experimental.get.vector.length

david-arm replied:
If #67934 is likely to land in the next week or so, I'm happy to hold off and base it off #67934. I can take a look at it, but I may need some help understanding how to do that. :)

Does your patch automatically solve this problem @fhahn, i.e. because it walks through every VPInstruction generated in the plan and so essentially there is no extra work needed?

david-arm replied:
I've updated this patch for reference anyway with the code changes and I've added some EVL check lines to the RISCV test. The test changes will be needed regardless.

fhahn replied:
> If #67934 is likely to land in the next week or so, I'm happy to hold off and base it off #67934. I can take a look at it, but I may need some help understanding how to do that. :)

Not sure about this week, but hopefully soon as all pending patches have landed.

> Does your patch automatically solve this problem @fhahn, i.e. because it walks through every VPInstruction generated in the plan and so essentially there is no extra work needed?

It won't solve it automatically; to solve it, `VPInstruction::computeCost` needs to be implemented for `ActiveLaneMask`, `GetVectorLength` and also `VPWidenCanonicalIVRecipe::compueCost`. Doing it that way would ensure we accurately cost what is actually code-gen'd (and avoid the issue with not knowing if a splat + add of step-vector will be needed as per my comment above)

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -5942,6 +5945,48 @@ InstructionCost LoopVectorizationCostModel::computePredInstDiscount(
+    // stepvector is loop invariant it will be hoisted out so we can ignore it.
+    // This just leaves us with an add and an icmp.
+    VectorType *VecTy = VectorType::get(IndTy, VF);
+    MaskCost = TTI.getArithmeticInstrCost(Instruction::Add, VecTy, CostKind);
```
Is this entirely accurate? We might need to also broadcast the canonical IV or there may be not increment needed, if we can re-use an existing widened IV. Unfortunately I am not sure if we can query this accurately at this point.

david-arm commented:
Hi @john-brawn-arm, whilst reviewing your PR #109289 I discovered I still have this patch that I created to add the cost of tail-folding to the loop. I think it may also fix the same issue you're looking at, albeit in a more accidental way.

At a reviewer request I held off doing further work on this PR until we moved over to the vplan cost model so that I could implement it differently, however it looks like we haven't moved over yet. @fhahn do you have any idea when we'll remove the assert that legacy cost == vplan cost?

Anyway, for the purpose of testing your issue @john-brawn-arm you can still try out this patch - I've simply commented out the assert that the legacy cost == vplan cost for now.

david-arm committed:
[LoopVectorize] Add cost of generating tail-folding mask to the loop

At the moment if we decide to enable tail-folding we do not include
the cost of generating the mask per VF. This can mean we make some
poor choices of VF, which is definitely true for SVE-enabled AArch64
targets where mask generation for fixed-width vectors is more
expensive than for scalable vectors.

I've added a VPInstruction::computeCost function to return the costs
of the ActiveLaneMask and ExplicitVectorLength operations.
Unfortunately, in order to prevent asserts firing I've also had to
duplicate the same code in the legacy cost model to make sure the
chosen VFs match up. I've wrapped this up in a ifndef NDEBUG for
now.

New tests added:

  Transforms/LoopVectorize/AArch64/sve-tail-folding-cost.ll
  Transforms/LoopVectorize/RISCV/tail-folding-cost.ll
